custom/sensors/PDP.java

```diff
@@ -36,17 +38,66 @@ public PDP(int ID) {
 		status2 = new CustomCAN("PDP STATUS 2", PDP.PDP_ID_STATUS_2 | ID);
 		status3 = new CustomCAN("PDP STATUS 2", PDP.PDP_ID_STATUS_3 | ID);
 		statusEnergy = new CustomCAN("PDP STATUS ENERGY", PDP.PDP_ID_STATUS_ENERGY | ID);
+		cachedVoltage = 11.5;
+		cachedCurrent = 0;
+		cachedChannelCurrents = new double[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+		cachedResistance = 80.0;
 	}
 
 	/**
 	 * PDP constructor
-	 * 
+	 *
 	 * This defaults to the PDP at ID 0.
 	 */
 	public PDP() {
 		this(0);
 	}
 
+	private void readStatus(int status) {
+		byte[] rawArray = null;
+		if (status == 1) {
+			rawArray = status1.read();
+		} else if (status == 2) {
+			rawArray = status1.read();
+		} else if (status == 3) {
+			readStatus3();
+		}
+		if (rawArray != null) {
+			double[] tempCurrents = new double[6];
+			tempCurrents[0] = ((rawArray[0] & 0xFF) << 2 | ((rawArray[1] & 0xC0) >> 6)) * 0.125;
+			tempCurrents[1] = (((rawArray[1] & 0x3F) << 4) | ((rawArray[2] & 0xF0) >> 4)) * 0.125;
+			tempCurrents[2] = (((rawArray[2] & 0x0F) << 6) | ((rawArray[3] & 0x3F) >> 2)) * 0.125;
+			tempCurrents[3] = (((rawArray[3] & 0xC0) << 8) | ((rawArray[4] & 0xFF))) * 0.125;
+			tempCurrents[4] = (((rawArray[5] & 0xFF) << 2) | ((rawArray[6] & 0xC0) >> 6)) * 0.125;
+			tempCurrents[5] = (((rawArray[6] & 0x3F) << 4) | ((rawArray[7] & 0xF0) >> 4)) * 0.125;
+			for (int i = 0; i < 6; i++) {
+				if (tempCurrents[i] < 128) { // deals with occasional issue with PDP reporting 1000+ amps (this is not a bug in this code, it was observed in PowerDistributionPanel as well
+					cachedChannelCurrents[i + (status - 1) * 6] = tempCurrents[i];
+				}
+			}
+			lastRead = System.currentTimeMillis();
+		}
+	}
+
+	private void readStatus3() { // Status 3 broken out for voltage and resistance
+		byte[] rawArray = status3.read();
+		if (rawArray != null) {
+			double[] tempCurrents = new double[4];
+			tempCurrents[0] = (((rawArray[0] & 0xFF) << 2) | ((rawArray[1] & 0xC0) >> 6)) * 0.125;
+			tempCurrents[1] = (((rawArray[1] & 0x3F) << 4) | ((rawArray[2] & 0xF0) >> 4)) * 0.125;
+			tempCurrents[2] = (((rawArray[2] & 0x0F) << 6) | ((rawArray[3] & 0x3F) >> 2)) * 0.125;
+			tempCurrents[3] = (((rawArray[3] & 0xC0) << 8) | ((rawArray[4] & 0xFF))) * 0.125;
+			for (int i = 0; i < 4; i++) {
+				if (tempCurrents[i] < 128) {
+					cachedChannelCurrents[i + 12] = tempCurrents[i];
+				}
+			}
+			cachedVoltage = (rawArray[6] & 0xFF) * 0.05 + 4.0;
+			cachedResistance = (rawArray[5] & 0xFF) / 1000.0; // in milliOhms
```

Comment:
Pedantic but perhaps the order of these should be flipped? (So it's ordered by the index)